File: haskell/Day1.hs
module Day1 where

solveP1 :: [Int] -> Int
solveP1 v = solve' v v
  where solve' (_ : as) [] = solve' as v
        solve' av@(a : _) (b : bs)
          | a + b == 2020 = a * b
          | otherwise = solve' av bs

solveP2 :: [Int] -> Int
solveP2 v = solve' v v v
  where solve' (_ : as) [] _ = solve' as v v
        solve' av (_ : bs) [] = solve' av bs v
        solve' av@(a : _) bv@(b : _) (c : cs)
          | a + b + c == 2020 = a * b * c
          | otherwise = solve' av bv cs

main :: IO ()
main = do
  inp <- lines <$> readFile "../day1.txt"
  let nums = map read inp
  print $ solveP1 nums
  print $ solveP2 nums

File: prolog/day1.pl
solveP1(A, B) :-
  member(X, A),
  member(Y, A),
  2020 #= X + Y,
  B is X * Y.

solveP2(A, B) :-
  member(X, A),
  member(Y, A),
  member(Z, A),
  2020 #= X + Y + Z,
  B is X * Y * Z.

read_nums(F, [H | T]) :-
  read_token(F, H),
  \+ at_end_of_stream(F),
  read_nums(F, T).
read_nums(_, []).

main :-
  open('../day1.txt', read, Str),
  read_nums(Str, Nums),
  solveP1(Nums, R1), !, print(R1), nl,
  solveP2(Nums, R2), !, print(R2), nl.

:- initialization(main).
